hazelcast-jet-kafka/src/main/java/com/hazelcast/jet/kafka/impl/WriteKafkaP.java

```diff
@@ -113,37 +210,150 @@ private void checkError() {
+    /**
+     * A simple class wrapping a KafkaProducer that ensures that
+     * `producer.initTransactions` is called exactly once before first
+     * `beginTransaction` call.
+     */
+    private static final class KafkaTransaction<K, V> implements TransactionalResource<KafkaTransactionId> {
+        private final KafkaProducer<K, V> producer;
+        private final ILogger logger;
+        private final KafkaTransactionId transactionId;
+        private boolean txnInitialized;
 
-        private final Properties properties;
-        private final Function<? super T, ? extends ProducerRecord<K, V>> toRecordFn;
+        private KafkaTransaction(KafkaTransactionId transactionId, Map<String, Object> properties, ILogger logger) {
+            this.transactionId = transactionId;
+            this.producer = new KafkaProducer<>(properties);
+            this.logger = logger;
+        }
 
-        private transient KafkaProducer<K, V> producer;
+        @Override
+        public KafkaTransactionId id() {
+            return transactionId;
+        }
 
-        public Supplier(Properties properties, Function<? super T, ? extends ProducerRecord<K, V>> toRecordFn) {
-            this.properties = properties;
-            this.toRecordFn = toRecordFn;
+        @Override
+        public void begin() {
+            if (!txnInitialized) {
+                LoggingUtil.logFine(logger, "initTransactions in begin %s", transactionId);
+                txnInitialized = true;
+                producer.initTransactions();
+                transactionId.updateProducerAndEpoch(producer);
+            }
+            producer.beginTransaction();
         }
 
         @Override
-        public void init(@Nonnull Context context) {
-            producer = new KafkaProducer<>(properties);
+        public boolean flush() {
+            producer.flush();
+            return true;
         }
 
-        @Override @Nonnull
-        public List<Processor> get(int count) {
-            return Stream.generate(() -> new WriteKafkaP<>(producer, toRecordFn))
-                         .limit(count)
-                         .collect(toList());
+        @Override
+        public void commit() {
+            if (transactionId != null) {
+                producer.commitTransaction();
+            }
+        }
+
+        @Override
+        public void rollback() {
+            if (transactionId != null) {
+                producer.abortTransaction();
+            }
+        }
+
+        @Override
+        public void release() {
+            producer.close();
+        }
+    }
+
+    public static class KafkaTransactionId implements TwoPhaseSnapshotCommitUtility.TransactionId, Serializable {
+        private final int processorIndex;
+        private long producerId = -1;
+        private short epoch = -1;
+        private final String kafkaId;
+        private final int hashCode;
+
+        KafkaTransactionId(long jobId, String jobName, @Nonnull String vertexId, int processorIndex,
```

Comment:
From the TDD:
>We include the jobName because two jobs submitted to two clusters at about the same time can have equal jobId (it's assigned by flake ID which contains mostly timestamp+sequence). 

But I admit it's very unlikely, but doesn't hurt...

Reply:
well, why do we care about uniqueness across clusters ?

Reply:
We should generate unique transaction IDs. We could use UUID, but we need it to be deterministic. So we try to put as many uniquifying factors as possible.

Reply:
For example, we could use "0" and "1" as transaction ID. But we will clash with a second Kafka sink writing to the same broker. So we add as many distinguishing elements as possible.